## Steps :
- Import libraries
- Import dataset
- EDA
- Min Max Scaling
- Encoding Technique
- Feature Engineering
- Feature Selection
- PPS Score

# Import libraries

In [19]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.preprocessing import MinMaxScaler

# Import dataset

In [2]:
df=pd.read_csv('adult_with_headers.csv')
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

# EDA

df.describe()

In [4]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education_num      int64
marital_status    object
occupation        object
relationship      object
race              object
sex               object
capital_gain       int64
capital_loss       int64
hours_per_week     int64
native_country    object
income            object
dtype: object

In [5]:
df.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

### Standard Scaler

In [6]:
numf=df.select_dtypes(include=['int64']).columns

In [7]:
scaler = StandardScaler()

df[numf]= scaler.fit_transform(df[numf])
df

age          workclass    fnlwgt    education  education_num  \
0      0.030671          State-gov -1.063611    Bachelors       1.134739   
1      0.837109   Self-emp-not-inc -1.008707    Bachelors       1.134739   
2     -0.042642            Private  0.245079      HS-grad      -0.420060   
3      1.057047            Private  0.425801         11th      -1.197459   
4     -0.775768            Private  1.408176    Bachelors       1.134739   
...         ...                ...       ...          ...            ...   
32556 -0.849080            Private  0.639741   Assoc-acdm       0.746039   
32557  0.103983            Private -0.335433      HS-grad      -0.420060   
32558  1.423610            Private -0.358777      HS-grad      -0.420060   
32559 -1.215643            Private  0.110960      HS-grad      -0.420060   
32560  0.983734       Self-emp-inc  0.929893      HS-grad      -0.420060   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male      0.148453      -0.21666       -0.035429   United-States   
1         Male     -0.145920      -0.21666       -2.222153   United-States   
2         Male     -0.145920      -0.21666       -0.035429   United-States   
3         Male     -0.145920      -0.21666       -0.035429   United-States   
4       Female     -0.145920      -0.21666       -0.035429            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female     -0.145920      -0.21666       -0.197409   United-States   
32557     Male     -0.145920      -0.21666       -0.035429   United-States   
32558   Female     -0.145920      -0.21666       -0.035429   United-States   
32559     Male     -0.145920      -0.21666       -1.655225   United-States   
32560   Female      1.888424      -0.21666       -0.035429   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

### Min Max Scaling

In [8]:
minmaxscaler = MinMaxScaler()

In [9]:
df[numf] = minmaxscaler.fit_transform(df[numf])
df

age          workclass    fnlwgt    education  education_num  \
0      0.301370          State-gov  0.044302    Bachelors       0.800000   
1      0.452055   Self-emp-not-inc  0.048238    Bachelors       0.800000   
2      0.287671            Private  0.138113      HS-grad       0.533333   
3      0.493151            Private  0.151068         11th       0.400000   
4      0.150685            Private  0.221488    Bachelors       0.800000   
...         ...                ...       ...          ...            ...   
32556  0.136986            Private  0.166404   Assoc-acdm       0.733333   
32557  0.315068            Private  0.096500      HS-grad       0.533333   
32558  0.561644            Private  0.094827      HS-grad       0.533333   
32559  0.068493            Private  0.128499      HS-grad       0.533333   
32560  0.479452       Self-emp-inc  0.187203      HS-grad       0.533333   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male      0.021740           0.0        0.397959   United-States   
1         Male      0.000000           0.0        0.122449   United-States   
2         Male      0.000000           0.0        0.397959   United-States   
3         Male      0.000000           0.0        0.397959   United-States   
4       Female      0.000000           0.0        0.397959            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female      0.000000           0.0        0.377551   United-States   
32557     Male      0.000000           0.0        0.397959   United-States   
32558   Female      0.000000           0.0        0.397959   United-States   
32559     Male      0.000000           0.0        0.193878   United-States   
32560   Female      0.150242           0.0        0.397959   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

•Discuss the scenarios where each scaling technique is preferred and why.

1. Standard Scaling (Z-score Normalization)
Preferred when:
Data is normally distributed or approximately normal.
Features have different units or scales (e.g., age in years, income in dollars).
Working with algorithms sensitive to distance or covariance (e.g., SVM, K-means, PCA).
You don’t want to distort relationships between features.
Why: Centers data around 0 with unit variance, ensuring features with larger magnitudes don't dominate.

2. Min-Max Scaling (Normalization)
Preferred when:
Data is not normally distributed or has outliers.
You need features on a fixed range (e.g., [0, 1]), especially for neural networks or algorithms like KNN.
Preserving relationships between features is important.
Why: Scales features to a fixed range, ensuring the model treats all features equally, but sensitive to outliers.

# Encoding Techniques

In [10]:
unique_counts = df[['age','workclass','fnlwgt','education','education_num','marital_status','occupation','relationship','race','sex','capital_gain','capital_loss','hours_per_week','native_country','income']].nunique()
unique_counts

age                  73
workclass             9
fnlwgt            21648
education            16
education_num        16
marital_status        7
occupation           15
relationship          6
race                  5
sex                   2
capital_gain        119
capital_loss         92
hours_per_week       94
native_country       42
income                2
dtype: int64

### One Hot Encoding

In [11]:
# Identify categorical columns with fewer than 5 unique categories
categorical_columns = df.select_dtypes(include=['object']).nunique()
columns_to_encode = categorical_columns[categorical_columns < 5].index

# Apply One-Hot Encoding to these columns
df_encoded = pd.get_dummies(df, columns=columns_to_encode, drop_first=False)

# Ensure output is in 0/1 format
df_encoded

age          workclass    fnlwgt    education  education_num  \
0      0.301370          State-gov  0.044302    Bachelors       0.800000   
1      0.452055   Self-emp-not-inc  0.048238    Bachelors       0.800000   
2      0.287671            Private  0.138113      HS-grad       0.533333   
3      0.493151            Private  0.151068         11th       0.400000   
4      0.150685            Private  0.221488    Bachelors       0.800000   
...         ...                ...       ...          ...            ...   
32556  0.136986            Private  0.166404   Assoc-acdm       0.733333   
32557  0.315068            Private  0.096500      HS-grad       0.533333   
32558  0.561644            Private  0.094827      HS-grad       0.533333   
32559  0.068493            Private  0.128499      HS-grad       0.533333   
32560  0.479452       Self-emp-inc  0.187203      HS-grad       0.533333   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

       capital_gain  capital_loss  hours_per_week  native_country  \
0          0.021740           0.0        0.397959   United-States   
1          0.000000           0.0        0.122449   United-States   
2          0.000000           0.0        0.397959   United-States   
3          0.000000           0.0        0.397959   United-States   
4          0.000000           0.0        0.397959            Cuba   
...             ...           ...             ...             ...   
32556      0.000000           0.0        0.377551   United-States   
32557      0.000000           0.0        0.397959   United-States   
32558      0.000000           0.0        0.397959   United-States   
32559      0.000000           0.0        0.193878   United-States   
32560      0.150242           0.0        0.397959   United-States   

       sex_ Female  sex_ Male  income_ <=50K  income_ >50K  
0            False       True           True         False  
1            False       True           True         False  
2            False       True           True         False  
3            False       True           True         False  
4             True      False           True         False  
...            ...        ...            ...           ...  
32556         True      False           True         False  
32557        False       True          False          True  
32558         True      False           True         False  
32559        False       True           True         False  
32560         True      False          False          True  

[32561 rows x 17 columns]

In [12]:
df_encoded.dtypes

age               float64
workclass          object
fnlwgt            float64
education          object
education_num     float64
marital_status     object
occupation         object
relationship       object
race               object
capital_gain      float64
capital_loss      float64
hours_per_week    float64
native_country     object
sex_ Female          bool
sex_ Male            bool
income_ <=50K        bool
income_ >50K         bool
dtype: object

### Label Encoder

In [13]:
# Identify categorical columns with more than 5 unique categories
categorical_columns = df.select_dtypes(include=['object']).nunique()
columns_to_encode = categorical_columns[categorical_columns > 5].index

# Apply Label Encoding to the selected columns
encoder = LabelEncoder()

# Label Encoding for each selected column
for column in columns_to_encode:
    df[column] = encoder.fit_transform(df[column])

# Display the DataFrame after Label Encoding
df

age  workclass    fnlwgt  education  education_num  \
0      0.301370          7  0.044302          9       0.800000   
1      0.452055          6  0.048238          9       0.800000   
2      0.287671          4  0.138113         11       0.533333   
3      0.493151          4  0.151068          1       0.400000   
4      0.150685          4  0.221488          9       0.800000   
...         ...        ...       ...        ...            ...   
32556  0.136986          4  0.166404          7       0.733333   
32557  0.315068          4  0.096500         11       0.533333   
32558  0.561644          4  0.094827         11       0.533333   
32559  0.068493          4  0.128499         11       0.533333   
32560  0.479452          5  0.187203         11       0.533333   

       marital_status  occupation  relationship    race      sex  \
0                   4           1             1   White     Male   
1                   2           4             0   White     Male   
2                   0           6             1   White     Male   
3                   2           6             0   Black     Male   
4                   2          10             5   Black   Female   
...               ...         ...           ...     ...      ...   
32556               2          13             5   White   Female   
32557               2           7             0   White     Male   
32558               6           1             4   White   Female   
32559               4           1             3   White     Male   
32560               2           4             5   White   Female   

       capital_gain  capital_loss  hours_per_week  native_country  income  
0          0.021740           0.0        0.397959              39   <=50K  
1          0.000000           0.0        0.122449              39   <=50K  
2          0.000000           0.0        0.397959              39   <=50K  
3          0.000000           0.0        0.397959              39   <=50K  
4          0.000000           0.0        0.397959               5   <=50K  
...             ...           ...             ...             ...     ...  
32556      0.000000           0.0        0.377551              39   <=50K  
32557      0.000000           0.0        0.397959              39    >50K  
32558      0.000000           0.0        0.397959              39   <=50K  
32559      0.000000           0.0        0.193878              39   <=50K  
32560      0.150242           0.0        0.397959              39    >50K  

[32561 rows x 15 columns]

### **One-Hot Encoding**  
**Pros**:
- No implicit ordering, suitable for nominal data.
- Works well for small to medium categories.
- Prevents models from assuming category order.

**Cons**:
- Increases dimensionality, leading to sparse matrices.
- Can be inefficient for high-cardinality features (many categories).

**Best For**:
- Nominal data (unordered categories).
- Small to medium-sized categorical features.

### **Label Encoding**  
**Pros**:
- Compact (uses one column).
- Efficient for ordinal data (ordered categories).
- Faster computation for large datasets.

**Cons**:
- Imposes an ordinal relationship on nominal data.
- Can mislead models by implying order where none exists.

**Best For**:
- Ordinal data (ordered categories).
- Large datasets with high-cardinality features.

# Feature Engineering

## (Education × Hours-Per-Week)

In [14]:
# Create the interaction feature
df['Education_Hours'] = df['education_num'] * df['hours_per_week']

# Preview the new feature
df[['education_num', 'hours_per_week', 'Education_Hours']].head()

education_num  hours_per_week  Education_Hours
0       0.800000        0.397959         0.318367
1       0.800000        0.122449         0.097959
2       0.533333        0.397959         0.212245
3       0.400000        0.397959         0.159184
4       0.800000        0.397959         0.318367

Rationale: If you have features like education level and hours worked per week, their interaction could capture the combined effect of education and work effort on income.

## (Income x Capital Ratio)

In [17]:
# Income-to-Capital Ratio (ICR)
# Add a small constant to avoid division by zero
df['ICR'] = (df['capital_gain'] - df['capital_loss']) / (df['fnlwgt'] + 1e-9)

# Display the first few rows with new features
df[['hours_per_week', 'relationship', 'capital_gain', 'capital_loss', 'fnlwgt', 'ICR']].head()

hours_per_week  relationship  capital_gain  capital_loss    fnlwgt  \
0        0.397959             1       0.02174           0.0  0.044302   
1        0.122449             0       0.00000           0.0  0.048238   
2        0.397959             1       0.00000           0.0  0.138113   
3        0.397959             0       0.00000           0.0  0.151068   
4        0.397959             5       0.00000           0.0  0.221488   

        ICR  
0  0.490729  
1  0.000000  
2  0.000000  
3  0.000000  
4  0.000000

Rationale: The *Income-to-Capital Ratio (ICR)* measures net capital gains relative to population representation (fnlwgt). It highlights financial security: positive values suggest wealthier individuals (higher income potential), while negative values indicate losses. Dividing by fnlwgt normalizes this metric, and a small constant prevents division errors.

### Skewnwss

In [20]:
# Identify numerical columns
numerical_columns = df.select_dtypes(include=[np.number]).columns
print("Numerical columns:", numerical_columns)

# Check skewness
for col in numerical_columns:
    skewness = df[col].skew()
    print(f"Skewness of {col}: {skewness}")

Numerical columns: Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'capital_gain',
       'capital_loss', 'hours_per_week', 'native_country', 'Education_Hours',
       'ICR'],
      dtype='object')
Skewness of age: 0.5587433694130481
Skewness of workclass: -0.7520240119541499
Skewness of fnlwgt: 1.4469800945789812
Skewness of education: -0.9340424374279611
Skewness of education_num: -0.31167586791022756
Skewness of marital_status: -0.01350813802823499
Skewness of occupation: 0.11458331643295894
Skewness of relationship: 0.7868177781306165
Skewness of capital_gain: 11.953847687699781
Skewness of capital_loss: 4.594629121679695
Skewness of hours_per_week: 0.22764253680450075
Skewness of native_country: -3.658303294717141
Skewness of Education_Hours: 0.727857086156734
Skewness of ICR: -3.085568991940199


The log transformation was applied to *capital_gain* because it had extremely high skewness (11.95). The transformation reduced the skewness, resulting in a more symmetric distribution, which improves model performance, interpretability, and statistical validity.

# Feature Selection

In [21]:
from sklearn.ensemble import IsolationForest

In [22]:
numfeats = df.select_dtypes(include=['int64', 'float64']).columns

In [23]:
# Applying Isolation Forest

isolation_forest = IsolationForest(contamination=0.01, random_state=42)
outliers = isolation_forest.fit_predict(df[numfeats])

In [24]:
df['outlier'] = outliers
print(df[['outlier']].value_counts())

outlier
 1         32235
-1           326
Name: count, dtype: int64


In [25]:
# Filter out outliers

data_no_outliers = df[df['outlier'] == 1]
data_no_outliers

age  workclass    fnlwgt  education  education_num  \
0      0.301370          7  0.044302          9       0.800000   
1      0.452055          6  0.048238          9       0.800000   
2      0.287671          4  0.138113         11       0.533333   
3      0.493151          4  0.151068          1       0.400000   
4      0.150685          4  0.221488          9       0.800000   
...         ...        ...       ...        ...            ...   
32556  0.136986          4  0.166404          7       0.733333   
32557  0.315068          4  0.096500         11       0.533333   
32558  0.561644          4  0.094827         11       0.533333   
32559  0.068493          4  0.128499         11       0.533333   
32560  0.479452          5  0.187203         11       0.533333   

       marital_status  occupation  relationship    race      sex  \
0                   4           1             1   White     Male   
1                   2           4             0   White     Male   
2                   0           6             1   White     Male   
3                   2           6             0   Black     Male   
4                   2          10             5   Black   Female   
...               ...         ...           ...     ...      ...   
32556               2          13             5   White   Female   
32557               2           7             0   White     Male   
32558               6           1             4   White   Female   
32559               4           1             3   White     Male   
32560               2           4             5   White   Female   

       capital_gain  capital_loss  hours_per_week  native_country  income  \
0          0.021740           0.0        0.397959              39   <=50K   
1          0.000000           0.0        0.122449              39   <=50K   
2          0.000000           0.0        0.397959              39   <=50K   
3          0.000000           0.0        0.397959              39   <=50K   
4          0.000000           0.0        0.397959               5   <=50K   
...             ...           ...             ...             ...     ...   
32556      0.000000           0.0        0.377551              39   <=50K   
32557      0.000000           0.0        0.397959              39    >50K   
32558      0.000000           0.0        0.397959              39   <=50K   
32559      0.000000           0.0        0.193878              39   <=50K   
32560      0.150242           0.0        0.397959              39    >50K   

       Education_Hours       ICR  outlier  
0             0.318367  0.490729        1  
1             0.097959  0.000000        1  
2             0.212245  0.000000        1  
3             0.159184  0.000000        1  
4             0.318367  0.000000        1  
...                ...       ...      ...  
32556         0.276871  0.000000        1  
32557         0.212245  0.000000        1  
32558         0.212245  0.000000        1  
32559         0.103401  0.000000        1  
32560         0.212245  0.802558        1  

[32235 rows x 18 columns]

In [26]:
# View outliers only

outliers_only = df[df['outlier'] == -1]
outliers_only

age  workclass    fnlwgt  education  education_num  \
52     0.410959          4  0.026861         14       0.933333   
157    0.739726          6  0.327310         15       0.600000   
222    1.000000          4  0.026799         11       0.533333   
272    0.452055          6  0.012475         12       0.866667   
573    0.273973          4  0.006912          8       0.666667   
...         ...        ...       ...        ...            ...   
32035  0.315068          4  0.376422         13       0.000000   
32090  0.671233          4  0.070098          9       0.800000   
32238  0.410959          4  0.191948         14       0.933333   
32443  0.013699          4  0.019761         11       0.533333   
32518  0.547945          2  0.066647         11       0.533333   

       marital_status  occupation  relationship    race      sex  \
52                  2          10             5   White   Female   
157                 5          12             4   Black     Male   
222                 4           8             1   Black     Male   
272                 2           5             0   White     Male   
573                 2           3             0   White     Male   
...               ...         ...           ...     ...      ...   
32035               2           8             0   White     Male   
32090               2           4             0   White     Male   
32238               2           4             0   White     Male   
32443               4          12             3   White   Female   
32518               2           3             0   White     Male   

       capital_gain  capital_loss  hours_per_week  native_country  income  \
52          0.00000      0.436639        0.602041              16    >50K   
157         0.00000      0.416896        0.010204              39   <=50K   
222         0.00000      0.506428        0.397959              39   <=50K   
272         0.02407      0.000000        0.989796              39   <=50K   
573         0.00000      0.453857        0.397959              39    >50K   
...             ...           ...             ...             ...     ...   
32035       0.00000      0.383838        0.397959              26   <=50K   
32090       1.00000      0.000000        0.551020               0    >50K   
32238       1.00000      0.000000        0.397959              39    >50K   
32443       0.00000      0.367769        0.193878              39   <=50K   
32518       1.00000      0.000000        0.397959              39    >50K   

       Education_Hours        ICR  outlier  
52            0.561905 -16.255781       -1  
157           0.006122  -1.273704       -1  
222           0.212245 -18.897452       -1  
272           0.857823   1.929524       -1  
573           0.265306 -65.658052       -1  
...                ...        ...      ...  
32035         0.000000  -1.019701       -1  
32090         0.440816  14.265838       -1  
32238         0.371429   5.209746       -1  
32443         0.103401 -18.611143       -1  
32518         0.212245  15.004484       -1  

[326 rows x 18 columns]

In [27]:
# Drop the 'outlier' column if it's no longer needed

data_no_outliers = data_no_outliers.drop(columns=['outlier'])

### PPS Score

In [ ]:
# Install the necessary library for PPS computation
!pip install ppscore --quiet

# Import the PPS library
import ppscore as pps

# Compute the PPS matrix for the dataset
pps_matrix = pps.matrix(cleaned_data)

# Display the top 10 feature relationships with the highest PPS
pps_matrix[['x', 'y', 'ppscore']].sort_values(by='ppscore', ascending=False).head(10)